### DISCUSSION TITLE
Welcome to PMD Discussions!

### DISCUSSION THREAD

**Discussion by adangel** [pmd/pmd#2998] [General]
We’re using Discussions as a place to connect with other members of our community. We hope that you:

* Ask questions you’re wondering about by creating a [new discussion](https://github.com/pmd/pmd/discussions/new)
* Searching for answers for already asked questions
* Share and develop ideas and features.
* Get help with a PR you're working on.
* Engage with other community members.

Discussions is one place to connect with the community. There is also our [Gitter Chatroom](https://gitter.im/pmd/pmd).

You can now also ask our [PMD Guru at gurubase](https://gurubase.io/g/pmd).

**Comment by aklmmap** [comment]
We use the following code to throw violations where a test class should use test methods. isTestMethodOrClass(node) is the predefined method used here to check whether a node here is test method or class.

it is working fine for test classes where the violations are being captured correctly.

Unfortunately this violation is flagging for triggers also where as it should only be flagged for test classes. Please find the snippet of the violation for a trigger in the attachment.

Code starts here:

public Object visit(final ASTMethod node, final Object data) {
// test methods should have @isTest annotation.
if (isTestMethodOrClass(node)) {
return data;
}
return checkForAssertStatements(node, data);
}

private Object checkForAssertStatements(final ASTMethod testMethod, final Object data) {
List<astmethodcallexpression> methodCallList = testMethod.findDescendantsOfType(ASTMethodCallExpression.class);
String assertMethodName;
for (ASTMethodCallExpression assertMethodCall : methodCallList) {
assertMethodName = assertMethodCall.getFullMethodName().toLowerCase(Locale.ROOT);
if ((!testMethod.getModifiers().isTestOrTestSetup() && (ASSERT_METHODS.contains(assertMethodName))) || (!testMethod.getModifiers().isTestOrTestSetup())) {
addViolationWithMessage(data, testMethod,
"Use @isTest or TestMethod or @TestSetup annotation for methods in Test Class.",
new Object[] { testMethod.getImage() });
return data;
}
}
return data;</astmethodcallexpression>
![trigger](https://user-images.githubusercontent.com/77477451/104684375-ea82ad00-571e-11eb-9467-3c20dd72a393.png)

**Reply by adangel** [reply]
@aklmmap I've created an own discussion for your question -> https://github.com/pmd/pmd/discussions/3059
